### PR TITLE
raftstore: make most ticks lazy

### PR DESCRIPTION
**Pull request by BusyJay**
## What have you changed? (mandatory)

This PR makes most ticks lazy except raft base tick. Raft base tick will
be implemented in another PR.

After this PR, down peers may not be reported in time. In the worse case
PD may need more than 5 minutes to know which node is down or up
again. /cc @disksing @nolouch 

## What are the type of the changes? (mandatory)

- New feature

## How has this PR been tested? (mandatory)

unit tests

## Does this PR affect documentation (docs) or release note? (mandatory)

Yes. But will be updated in the next PR.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#4591, tikv/rfcs#23

### CODE REVIEW

File: src/raftstore/store/fsm/peer.rs [hunk 1]
```diff
@@ -2380,6 +2413,7 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         }
         self.fsm.peer.size_diff_hint = 0;
         self.fsm.peer.compaction_declined_bytes = 0;
+        self.register_split_region_check_tick();
```
Comment (Connor1996) [inline src/raftstore/store/fsm/peer.rs]: seems it is unnecessary

File: src/raftstore/store/fsm/peer.rs [hunk 2]
```diff
@@ -2558,6 +2599,10 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
             return;
         }
 
+        if self.fsm.peer.is_leader() {
```
Comment (Connor1996) [inline src/raftstore/store/fsm/peer.rs]: please add some comments to explain why we should do that

File: src/raftstore/store/fsm/peer.rs [hunk 3]
```diff
@@ -272,6 +271,11 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
                 PeerMsg::SignificantMsg(msg) => self.on_significant_msg(msg),
                 PeerMsg::CasualMessage(msg) => self.on_casual_msg(msg),
                 PeerMsg::Start => self.start(),
+                PeerMsg::HeartbeatPd => {
+                    if self.fsm.peer.is_leader() {
+                        self.register_pd_heartbeat_tick()
```
Comment (overvenus) [inline src/raftstore/store/fsm/peer.rs]: Why not heartbeat pd immediately?

File: src/raftstore/store/fsm/store.rs [hunk 4]
```diff
@@ -1022,6 +1022,14 @@ impl RaftBatchSystem {
         self.apply_system
             .schedule_all(region_peers.iter().map(|pair| pair.1.get_peer()));
 
+        let router = Mutex::new(self.router.clone());
+        pd_client.handle_reconnect(move || {
+            router
+                .lock()
+                .unwrap()
+                .broadcast_normal(|| PeerMsg::HeartbeatPd);
```
Comment (overvenus) [inline src/raftstore/store/fsm/store.rs]: How about `PeerMsg::Tick(HeartbeatPd)`?

File: src/raftstore/store/peer.rs [hunk 5]
```diff
@@ -703,8 +707,9 @@ impl Peer {
     }
 
     /// Collects all down peers.
-    pub fn collect_down_peers(&self, max_duration: Duration) -> Vec<PeerStats> {
+    pub fn collect_down_peers(&mut self, max_duration: Duration) -> Vec<PeerStats> {
         let mut down_peers = Vec::new();
+        let mut down_peer_ids = Vec::new();
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: What about using `self.down_peer_ids`?

File: src/raftstore/store/peer.rs [hunk 6]
```diff
@@ -759,13 +766,16 @@ impl Peer {
...
+    ///
+    /// A peer can become pending or down if it has not responded for a
+    /// long time. If it becomes normal again, PD need to be notified.
     pub fn any_new_peer_catch_up(&mut self, peer_id: u64) -> bool {
-        if self.peers_start_pending_time.is_empty() {
+        if self.peers_start_pending_time.is_empty() && self.down_peer_ids.is_empty() {
             return false;
         }
         if !self.is_leader() {
+            self.down_peer_ids = vec![];
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: Why not `clear`?

File: src/raftstore/store/fsm/peer.rs [hunk 7]
```diff
@@ -567,7 +583,15 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         if let Some(p) = self.fsm.peer.take_apply_proposals() {
             proposals.push(p);
         }
-        self.fsm.peer.handle_raft_ready_append(self.ctx);
+        let res = self.fsm.peer.handle_raft_ready_append(self.ctx);
+        if let Some(r) = res {
+            self.on_role_changed(&r.0);
```
Comment (overvenus) [inline src/raftstore/store/fsm/peer.rs]: Looks like it has been handled in `handle_raft_ready_append ` already?

File: src/raftstore/store/fsm/peer.rs [hunk 8]
```diff
@@ -728,6 +752,7 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
                     res.merged,
                     &res.metrics,
                 );
+                self.register_pd_heartbeat_tick();
```
Comment (overvenus) [inline src/raftstore/store/fsm/peer.rs]: Could you add some comments about why we need to heartbeat pd after applying something?

File: src/raftstore/store/fsm/peer.rs (hunk 7 above)
Comment (BusyJay) [reply src/raftstore/store/fsm/peer.rs]: Different layers execute different commands.

File: src/raftstore/store/fsm/peer.rs (hunk 1 above)
Comment (BusyJay) [reply src/raftstore/store/fsm/peer.rs]: Split regions may still need to be split again.

File: src/raftstore/store/fsm/store.rs (hunk 4 above)
Comment (BusyJay) [reply src/raftstore/store/fsm/store.rs]: Do you mean replace the message with PeerTicks? I think it's a good idea to separate automatically tick from custom messages.

File: src/raftstore/store/peer.rs (hunk 5 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: Lifetime problems. Can't mutate it after L713.

File: src/raftstore/store/peer.rs (hunk 6 above)
Comment (BusyJay) [reply src/raftstore/store/peer.rs]: Release memory.

File: src/raftstore/store/fsm/peer.rs (hunk 1 above)
Comment (Connor1996) [reply src/raftstore/store/fsm/peer.rs]: it will register tick in `on_ready_split_region` to split again, so need here

Comment (BusyJay) [reply src/raftstore/store/fsm/peer.rs]: Oh, forget about what I have said. Here is just to make it robust, as the task may be lost for some reason.

File: src/raftstore/store/fsm/peer.rs (hunk 3 above)
Comment (BusyJay) [reply src/raftstore/store/fsm/peer.rs]: To reduce heartbeats. There may be pd heartbeats already. And in the past, it also take at most 10 seconds to heartbeat the new PD.

**Comment by BusyJay** [comment]
PTAL

**Pull request merged**